traits/ctraits.c

```diff
@@ -3339,7 +3339,7 @@ validate_trait_integer(
 */
 
 static PyObject *
-as_float(PyObject *value)
+validate_float(PyObject *value)
```

Comment:
IIUC, the internals of `validate_float` e.g. `PyFloat_AsDouble` and others handle `__index__` correctly, which is why this is the right approach - which is also what you meant by

> difficulty of expressing the condition to be accepted as a float in pure Python

Is that correct?

Reply:
Yes! Moreover, the Python behaviour changes from version to version, and we want to match Python's behaviour regardless of Python version. The easiest way to do that is to use Python directly rather than trying to encode Python's logic in our own code (which would then mean having to update that code whenever Python's behaviour changes).

Related: https://bugs.python.org/issue40801